### PR TITLE
htlcswitch: fix alignment of the packetQueue's fields for 32-bit systems

### LINKED ISSUE #505: Panic: runtime error: invalid memory address or nil pointer dereference
Hello, 
I setup lnd on a raspbberyPi 3 (running in neutrino mode) and after 1 day of running I got the below crash in the logs:

> 2017-12-20 14:52:05.494 [INF] NTFN: New spend subscription: utxo=64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0
> 2017-12-20 14:52:05.495 [INF] LNWL: Close observer for ChannelPoint(64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0) active
> 2017-12-20 14:52:05.531 [DBG] BRAR: New contract detected, launching breachObserver
> 2017-12-20 14:52:05.531 [INF] PEER: New channel active ChannelPoint(64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0) with peerId(3)
> 2017-12-20 14:52:05.531 [DBG] BRAR: Breach observer for ChannelPoint(64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0) started
> 2017-12-20 14:52:05.532 [INF] NTFN: New block epoch subscription
> 2017-12-20 14:52:05.532 [INF] HSWC: ChannelLink(64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0) is starting
> 2017-12-20 14:52:05.532 [INF] HSWC: Added channel link with chan_id=3d9207233d6cd82996a8248b1f0b63023efaafdffc11eb7fd6a7ff86037fdb64, short_chan_id=((lnwire.ShortChannelID) 1255466:79:0
> )
> 2017-12-20 14:52:05.536 [INF] HSWC: ChannelLink(64db7f0386ffa7d67feb11fcdfaffa3e02630b1f8b24a89629d86c3d2307923d:0) has exited
> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x4 pc=0x134ac]
> 
> goroutine 260616 [running]:
> sync/atomic.loadUint64(0x253ea7c4, 0x0, 0x27a2a094)
>         /usr/local/go/src/sync/atomic/64bit_arm.go:10 +0x3c
> github.com/lightningnetwork/lnd/htlcswitch.(*packetQueue).TotalHtlcAmount(0x253ea7c0, 0x1dcd6500, 0x0)
>         /home/pi/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/queue.go:197 +0x24
> github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).Bandwidth(0x12562500, 0x25694624, 0x12562500)
>         /home/pi/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:1060 +0x40
> github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).htlcManager(0x12562500)
>         /home/pi/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:362 +0x5c
> created by github.com/lightningnetwork/lnd/htlcswitch.(*channelLink).Start
>         /home/pi/gocode/src/github.com/lightningnetwork/lnd/htlcswitch/link.go:267 +0x104

### PR DESCRIPTION
**Pull request by Roasbeef**
In this commit, we fix an existing issue that would cause lnd to panic
on 32-bit systems. Within the packetQueue we utilize atomics heavily.
However, it's the caller's job to ensure 64-bit alignment of 64-bit words
accessed atomically. This is documented within the sync/atomic package
as a set of known bugs.

The old alignment of this struct was:

⛰  structlayout github.com/lightningnetwork/lnd/htlcswitch packetQueue
packetQueue.queueLen int32: 0-4 (size 4, align 4)
padding: 4-8 (size 4, align 0)
packetQueue.totalHtlcAmt int64: 8-16 (size 8, align 8)
packetQueue.queueCond *sync.Cond: 16-24 (size 8, align 8)
packetQueue.queueMtx.state int32: 24-28 (size 4, align 4)
packetQueue.queueMtx.sema uint32: 28-32 (size 4, align 4)
packetQueue.queue []*github.com/lightningnetwork/lnd/htlcswitch.htlcPacket: 32-56 (size 24, align 8)
packetQueue.outgoingPkts chan *github.com/lightningnetwork/lnd/htlcswitch.htlcPacket: 56-64 (size 8, align 8)
packetQueue.freeSlots chan struct{}: 64-72 (size 8, align 8)
packetQueue.wg.noCopy sync.noCopy: 72-72 (size 0, align 1)
packetQueue.wg.state1 [12]byte: 72-84 (size 12, align 1)
packetQueue.wg.sema uint32: 84-88 (size 4, align 4)
packetQueue.quit chan struct{}: 88-96 (size 8, align 8)

After this commit, the new alignment of this sturct is:

⛰  structlayout  -json github.com/lightningnetwork/lnd/htlcswitch packetQueue | structlayout-optimize
packetQueue.queue []*github.com/lightningnetwork/lnd/htlcswitch.htlcPacket: 0-24 (size 24, align 8)
packetQueue.wg struct: 24-40 (size 16, align 8)
packetQueue.freeSlots chan struct{}: 40-48 (size 8, align 8)
packetQueue.queueCond *sync.Cond: 48-56 (size 8, align 8)
packetQueue.queueMtx struct: 56-64 (size 8, align 8)
packetQueue.outgoingPkts chan *github.com/lightningnetwork/lnd/htlcswitch.htlcPacket: 64-72 (size 8, align 8)
packetQueue.totalHtlcAmt int64: 72-80 (size 8, align 8)
packetQueue.quit chan struct{}: 80-88 (size 8, align 8)
packetQueue.queueLen int32: 88-92 (size 4, align 8)
padding: 92-96 (size 4, align 0)

Fixes #505, and #463.

### CODE REVIEW

**Comment by creslinux** [comment]
I have a Raspberry Pi3 with 32bit goloang that was consistently failing to start lnd due to the struct not being aligned. The Pi worked for a few days prior. I have applied the patch to my Pi3.

My Pi3 now fails at a different stage that may not be related. 
```
2017-12-21 12:05:28.262 [TRC] FNDG: Funding manager running
unable to read channel for chain_hash=43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000, node_key=02f1da524a70afd8de6019e2367b47d8d41a623aa3594f55d0785fe1b047c853bc: unable to read channel data for chan_point=105b920b4a61f1bbee93cf1c37f756a870f9b13bb4df60cd21cfa0c7abca6534:0: unable to fetch chan info: invalid magic in compressed pubkey string: 241 
```
I would temper this test result with the disclaimer I've used github for only 3 days and have zero knowledge of the golang language. I applied the patch as shown in this post - which may be incorrect. I did manually check the contents of the queue.go file prior to glide install and can confirm it was the verison with the PR changes. 

I have included in this comment 

- My patch method to apply queue.go. 
- My lnd startup log. 
- The excerpt from btcd logfile when fails. 

Apply patch method:
```
cd $GOPATH/src/github.com/lightningnetwork/lnd
git pull
cd htlcswitch/
rm queue.go
wget https://raw.githubusercontent.com/Roasbeef/lnd/38673eb8e9a2a961b38249a6bacc790afd20aae7/htlcswitch/queue.go
cd $GOPATH/src/github.com/lightningnetwork/lnd
glide install
go install . ./cmd/... 
cd 
lnd --bitcoin.active --bitcoin.testnet
```
My lnd start up log:
```
lnd --bitcoin.active --bitcoin.testnet
2017-12-21 12:04:58.023 [INF] LTND: Version 0.3.0-alpha
2017-12-21 12:04:58.025 [INF] CHDB: Checking for schema update: latest_version=0, db_version=0
2017-12-21 12:04:58.051 [INF] RPCS: password gRPC proxy started at localhost:8080
2017-12-21 12:04:58.051 [INF] RPCS: password RPC server listening on 127.0.0.1:10009
2017-12-21 12:04:58.051 [INF] LTND: Waiting for wallet encryption password. Use `lncli create` to create wallet, or `lncli unlock` to unlock already created wallet.
2017-12-21 12:05:11.500 [INF] LNWL: Opened wallet
2017-12-21 12:05:11.501 [INF] LTND: Primary chain is set to: bitcoin
2017-12-21 12:05:11.502 [INF] LTND: Initializing btcd backed fee estimator
2017-12-21 12:05:19.906 [INF] LNWL: Opened wallet
2017-12-21 12:05:28.186 [INF] LNWL: The wallet has been unlocked without a time limit
2017-12-21 12:05:28.230 [INF] LTND: LightningWallet opened
2017-12-21 12:05:28.233 [DBG] LTND: Adding debug invoice (*channeldb.Invoice)(0x111f8310)({
 Memo: ([]uint8) <nil>,
 Receipt: ([]uint8) <nil>,
 PaymentRequest: ([]uint8) <nil>,
 CreationDate: (time.Time) 2017-12-21 12:05:28.233147366 +0200 EET,
 SettleDate: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
 Terms: (channeldb.ContractTerm) {
  PaymentPreimage: ([32]uint8) (len=32 cap=32) {
   00000000  01 01 01 01 01 01 01 01  01 01 01 01 01 01 01 01  |................|
   00000010  01 01 01 01 01 01 01 01  01 01 01 01 01 01 01 01  |................|
  },
  Value: (lnwire.MilliSatoshi) 100000000000000 mSAT,
  Settled: (bool) false
 }
})

2017-12-21 12:05:28.233 [DBG] SRVR: Debug HTLC invoice inserted, preimage=0101010101010101010101010101010101010101010101010101010101010101, hash=72cd6e8422c407fb6d098690f1130b7ded7ec2f7f5e1d30bd9d521f015363793
2017-12-21 12:05:28.238 [INF] LNWL: Started rescan from block 00000000000a59dfbc24606544ff4fba68be98cf7dc651b65f5ade26c16303cd (height 1255229) for 17 addresses
2017-12-21 12:05:28.262 [TRC] FNDG: Funding manager running
unable to read channel for chain_hash=43497fd7f826957108f4a30fd9cec3aeba79972084e90ead01ea330900000000, node_key=02f1da524a70afd8de6019e2367b47d8d41a623aa3594f55d0785fe1b047c853bc: unable to read channel data for chan_point=105b920b4a61f1bbee93cf1c37f756a870f9b13bb4df60cd21cfa0c7abca6534:0: unable to fetch chan info: invalid magic in compressed pubkey string: 241
```

btcd log
```
2017-12-21 12:05:12.042 [INF] RPCS: New websocket client [::1]:57546
2017-12-21 12:05:20.902 [INF] RPCS: New websocket client [::1]:57548
2017-12-21 12:05:28.240 [INF] RPCS: Beginning rescan for 17 addresses
2017-12-21 12:05:28.273 [ERR] RPCS: Websocket receive error from [::1]:57548: websocket: close 1006 unexpected EOF
2017-12-21 12:05:28.273 [ERR] RPCS: Websocket receive error from [::1]:57546: websocket: close 1006 unexpected EOF
2017-12-21 12:05:28.273 [INF] RPCS: Disconnected websocket client [::1]:57548
2017-12-21 12:05:28.273 [INF] RPCS: Disconnected websocket client [::1]:57546

```

**Comment by Roasbeef** [comment]
@creslinux thanks for giving this PR a spin!

The error you encountered isn't related to this PR. Rather you were on an older version of `lnd` before a recent breaking database change.

**Comment by creslinux** [comment]
@roasbeef thanks for the patch and feedback )

If I run clean install of lnd with the queue.go patch; removing the  src tree and .lnd directories first would this get me past the database incompatibility.  Or do I need to alter btcd also.

Thanks

**Comment by halseth** [comment]
@creslinux Cleaning the `.lnd` dir should suffice!

File: htlcswitch/queue.go [hunk 1]
```diff
@@ -17,34 +17,36 @@ import (
...
 // the packetQueue to know when new items have been added to the queue.
 type packetQueue struct {
-	// queueLen is an internal counter that reflects the size of the queue
-	// at any given instance. This value is intended to be use atomically
-	// as this value is used by internal methods to obtain the length of
-	// the queue w/o grabbing the main lock. This allows callers to avoid a
-	// deadlock situation where the main goroutine is attempting a send
-	// with the lock held.
-	queueLen int32
+	queue []*htlcPacket
```
Comment (halseth) [inline htlcswitch/queue.go]: Should we add any comments explaining why/warning that the order of these are important for alignment?

Comment (Roasbeef) [reply htlcswitch/queue.go]: Eh, I'm not sure it's necessary. Typically in the codebase we group attributes of a struct logically (things that are utilized together are laid out close together in the struct). The new version of the structure _mostly_ adheres to this, just needed to swap some things to get rid of bad alignment. 

Usage of atomics within the codebase is _definitely_ something that should be high lighted in our code contribution guidelines. Or just a new document going over common pitfalls (like referencing a range variable in a closure, etc).

**Comment by creslinux** [comment]
@halseth Great news. I'll restart a build with a cleaned .lnd and let the thread know the result in an hour.

**Comment by creslinux** [comment]
I have a Raspberry Pi3 with 32bit golang that was consistently failing to start lnd due to the struct not being aligned. The Pi worked for a few days prior. I have applied the patch to my Pi3.

**My Pi3 with 32bit goland lnd daemon now starts without error and I am unable to trigger the error previously seen.** 

1. I have created new channels from my Pi to the faucet and also to an Elcair android wallet. 
2. I have successfully purchased coffees from starblocks through the Pi3 lnd node 
3. I have restarted lnd 3 times without err and made payment from eclair through the lnd node on my Pi3 to starblocks after

**Comment by Roasbeef** [comment]
Nice! Long live the raspi!

**Pull request merged**